roaring/roaring.go

```diff
@@ -705,93 +674,98 @@ func (b *Bitmap) unionInPlace(others ...*Bitmap) {
-			}
+			iKey, iContainer := iIter.iter.Value()
+			expectedN := int64(0)
 
-			// There was more than one container across the bitmaps with key iKey
-			// so we need to calculate a union.
+			// determine whether we have a target to union into.
+			tContainer := target.Containers.Get(iKey)
+			// if the target's full, short-circuit out.
+			if tContainer != nil {
+				if tContainer.n == maxContainerVal+1 {
+					bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
+					continue
+				}
+				expectedN = int64(tContainer.n)
+			}
+			// Check i and later iters for any max-range containers, and
+			// find out how many there are.
+			summaryStats := bitmapIters[i:].calculateSummaryStats(iKey)
 			if summaryStats.hasMaxRange {
 				// One (or more) of the containers represented the maximum possible
 				// range that a container can store, so instead of calculating a
 				// union we can generate an RLE container that represents the entire
 				// range.
-				container := &Container{
+				tContainer = &Container{
 					runs:          []interval16{{start: 0, last: maxContainerVal}},
 					containerType: containerRun,
 					n:             maxContainerVal + 1,
 				}
-				target.Containers.Put(iKey, container)
+				target.Containers.Put(iKey, tContainer)
 				bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
 				continue
 			}
-
-			// Use a bitmap container for the target bitmap, and union everything
-			// into it.
-			//
-			// TODO(rartoul): Add another conditional case for n < ArrayMaxSize
-			// (or some fraction of that value) to avoid allocating expensive
-			// bitmaps when unioning many low-density array containers, but this
-			// will require writing a union in place algorithm for an array container
-			// that accepts multiple different containers to union into it for
-			// efficiency.
-			container := target.Containers.Get(iKey)
-			// If target already has a bitmap container for iKey then we can reuse that,
-			// otherwise we have to allocate a new one or convert the existing container
-			// into a bitmap container.
-			if container == nil {
-				buf := make([]uint64, bitmapN)
-				ob := buf[:bitmapN]
-				container = &Container{
-					bitmap:        ob,
-					n:             0,
-					containerType: containerBitmap,
+			expectedN += summaryStats.n
+			var itersToUnion handledIters
+			if tContainer == nil {
+				if summaryStats.c == 1 {
+					// There's no target and we have only one container, we
+					// can just clone it instead of unioning
+					statsHit("unionInPlace/reuse")
+					target.Containers.Put(iKey, iContainer.Clone())
+					bitmapIters[i].handled = true
+					continue
+				}
+				// If we expect enough items that we'll want to convert to bitmap,
+				// and the container we already found isn't a bitmap, make a bitmap
+				// and union everything into that. Otherwise clone the first
+				// container and union everything else into it.
+				if expectedN >= 512 && iContainer.containerType != containerBitmap {
```

Comment:
I thought max array size was like 4096? Maybe I'm misremembering

Reply:
2048. But! It turns out that, if you do an array/array in-place merge of >500 items, it's *faster* to do this -- doing the conversion only if the expected N is that high results in a factor of two or more performance regression for use cases where we get about 1k items in a container.

Reply:
For the record, array max size is 4096, run max size is 2048.

Reply:
... That is true, and also obvious, and now I'm not sure how I managed to get confused about it. Anyway, cutover is way lower than the max size for an array, just because the array merge is not very cheap. I'm not sure whether there's an easy way to make it cheaper.

Reply:
We could probably tweak ArrayMaxSize down to improve query performance at the cost of increased storage. Whether or not this would be worthwhile probably depends on the particular use case and workload. I'm inclined to believe we have easier and more general wins elsewhere though before we start tuning at that level.

Reply:
Yeah. Might be fun to benchmark, but there's a lot of additional tradeoffs, etc., so.